ast/src/parsed/display.rs

```diff
@@ -587,6 +595,84 @@ fn format_list<L: IntoIterator<Item = I>, I: Display>(list: L) -> String {
     format!("{}", list.into_iter().format(", "))
 }
 
+impl<E: Display> Expression<E> {
+    pub fn precedence(&self) -> Option<ExpressionPrecedence> {
+        match self {
+            Expression::UnaryOperation(op, _) => Some(op.precedence()),
+            Expression::BinaryOperation(_, op, _) => Some(op.precedence()),
+            _ => None,
+        }
+    }
+
+    pub fn is_postfix(&self) -> bool {
+        match self {
+            Expression::UnaryOperation(op, _) => !op.is_prefix(),
+            Expression::FunctionCall(_) | Expression::IndexAccess(_) => true,
+            _ => false,
+        }
+    }
+
+    pub fn format_unary_operation(
+        &self,
+        op: &UnaryOperator,
+        exp: &Expression<E>,
+        f: &mut Formatter<'_>,
+    ) -> Result {
+        let exp_string = match (self.precedence(), exp.precedence()) {
+            (Some(precedence), Some(inner_precedence)) if precedence < inner_precedence => {
+                format!("({exp})")
+            }
+            _ => {
+                format!("{exp}")
+            }
+        };
+
+        if op.is_prefix() {
+            write!(f, "{}{}", op, exp_string)
+        } else {
+            write!(f, "{}{}", exp_string, op)
+        }
+    }
+
+    pub fn format_binary_operation(
+        left: &Expression<E>,
+        op: &BinaryOperator,
+        right: &Expression<E>,
+        f: &mut Formatter<'_>,
+    ) -> Result {
+        let use_left_parentheses = match left.precedence() {
+            Some(left_precedence) => {
+                left_precedence > op.precedence()
+                    || (left_precedence == op.precedence()
+                        && op.associativity() != BinaryOperatorAssociativity::Left)
+            }
+            None => false,
+        };
+
+        let use_right_parentheses = match right.precedence() {
+            Some(right_precedence) => {
+                right_precedence > op.precedence()
+                    || (right_precedence == op.precedence()
+                        && op.associativity() != BinaryOperatorAssociativity::Right)
+            }
+            None => false,
+        };
+
+        let left_string = if use_left_parentheses {
+            format!("({})", left)
```

Comment:
same here and in many other places:
```suggestion
            format!("({left})")
```